src/csharp/Microsoft.Spark/ML/Feature/Base.cs

```diff
@@ -0,0 +1,332 @@
+
+    /// <summary>
+    /// Abstract Class for estimators that fit models to data.
+    /// </summary>
+    /// <typeparam name="M"/>
+    public abstract class ScalaEstimator<M> : ScalaPipelineStage, Estimator<M> where M : ScalaModel<M>
+    {
+        internal ScalaEstimator(string className) : base(className)
+        {
+        }
+
+        internal ScalaEstimator(string className, string uid) : base(className, uid)
+        {
+        }
+
+        internal ScalaEstimator(JvmObjectReference jvmObject) : base(jvmObject)
+        {
+        }
+
+        /// <summary>
+        /// Fits a model to the input data.
+        /// </summary>
+        /// <param name="dataset">input dataset.</param>
+        /// <returns>fitted model</returns>
+        public virtual M Fit(DataFrame dataset) =>
+            WrapAsType<M>((JvmObjectReference)Reference.Invoke("fit", dataset));
+    }
+
+    /// <summary>
+    /// A helper interface for ScalaModel, so that when we have an array of ScalaModels
+    /// with different type params, we can hold all of them with Model&lt;object&gt;.
+    /// </summary>
+    public interface Model<out M>
+    {
+        bool HasParent();
+    }
+
+    /// <summary>
+    /// A fitted model, i.e., a Transformer produced by an Estimator.
+    /// </summary>
+    /// <typeparam name="M">
+    /// Model Type.
+    /// </typeparam>
+    public abstract class ScalaModel<M> : ScalaTransformer, Model<M> where M : ScalaModel<M>
+    {
+        internal ScalaModel(string className) : base(className)
+        {
+        }
+
+        internal ScalaModel(string className, string uid) : base(className, uid)
+        {
+        }
+
+        internal ScalaModel(JvmObjectReference jvmObject) : base(jvmObject)
+        {
+        }
+
+        /// <summary>
+        /// Sets the parent of this model (Java API).
+        /// </summary>
+        /// <returns>type parameter M</returns>
+        public M SetParent(ScalaEstimator<M> parent) =>
+            WrapAsType<M>((JvmObjectReference)Reference.Invoke("setParent", parent));
+
+        /// <summary>
+        /// Indicates whether this Model has a corresponding parent.
+        /// </summary>
+        /// <returns>bool</returns>
+        public bool HasParent() =>
+            (bool)Reference.Invoke("hasParent");
+    }
+
+    /// <summary>
+    /// <see cref="ScalaEvaluator"/> Abstract Class for evaluators that compute metrics from predictions.
+    /// </summary>
+    public abstract class ScalaEvaluator : Params
+    {
+        internal ScalaEvaluator(string className) : base(className)
+        {
+        }
+
+        internal ScalaEvaluator(string className, string uid) : base(className, uid)
+        {
+        }
+
+        internal ScalaEvaluator(JvmObjectReference jvmObject) : base(jvmObject)
+        {
+        }
+
+        /// <summary>
+        /// Evaluates model output and returns a scalar metric.
+        /// The value of isLargerBetter specifies whether larger values are better.
```

Comment:
I'm not sure I understand this description. is `isLargerBetter` an example of a scalar metric that is returned based on the input Dataframe?

Reply:
This is the exact description of original apache spark.ml.evaluation.Evalutor's method. isLargerBetter is a boolean, that indicates whether the metric returned (for example accuracy) is larger the better.